nexus/src/db/model/ip_pool.rs

```diff
@@ -0,0 +1,127 @@
+use diesel::Selectable;
+use ipnetwork::IpNetwork;
+use omicron_common::api::external;
+use std::net::IpAddr;
+use uuid::Uuid;
+
+/// An IP Pool is a collection of IP addresses external to the rack.
+#[derive(Queryable, Insertable, Selectable, Clone, Debug, Resource)]
+#[diesel(table_name = ip_pool)]
+pub struct IpPool {
+    #[diesel(embed)]
+    pub identity: IpPoolIdentity,
+
+    /// Child resource generation number, for optimistic concurrency control of
+    /// the contained ranges.
+    pub rcgen: i64,
+}
+
+impl IpPool {
+    pub fn new(pool_identity: &external::IdentityMetadataCreateParams) -> Self {
+        Self {
+            identity: IpPoolIdentity::new(
+                Uuid::new_v4(),
+                pool_identity.clone(),
+            ),
+            rcgen: 0,
+        }
+    }
+}
+
+/// A set of updates to an IP Pool
+#[derive(AsChangeset)]
+#[diesel(table_name = ip_pool)]
+pub struct IpPoolUpdate {
+    pub name: Option<Name>,
+    pub description: Option<String>,
+    pub time_modified: DateTime<Utc>,
+}
+
+impl From<params::IpPoolUpdate> for IpPoolUpdate {
+    fn from(params: params::IpPoolUpdate) -> Self {
+        Self {
+            name: params.identity.name.map(|n| n.into()),
+            description: params.identity.description,
+            time_modified: Utc::now(),
+        }
+    }
+}
+
+/// A range of IP addresses for an IP Pool.
+#[derive(Queryable, Insertable, Selectable, Clone, Debug)]
+#[diesel(table_name = ip_pool_range)]
+pub struct IpPoolRange {
+    pub id: Uuid,
+    pub time_created: DateTime<Utc>,
+    pub time_modified: DateTime<Utc>,
+    pub time_deleted: Option<DateTime<Utc>>,
+    pub first_address: IpNetwork,
+    pub last_address: IpNetwork,
+    pub ip_pool_id: Uuid,
+}
+
+impl IpPoolRange {
+    pub fn new(range: &external::IpRange, ip_pool_id: Uuid) -> Self {
+        let now = Utc::now();
+        let first_address = range.first_address();
+        let last_address = range.last_address();
+        assert!(
+            last_address >= first_address,
+            "Address ranges must be non-decreasing"
+        );
+        Self {
+            id: Uuid::new_v4(),
+            time_created: now,
+            time_modified: now,
+            time_deleted: None,
+            first_address: IpNetwork::from(range.first_address()),
+            last_address: IpNetwork::from(range.last_address()),
+            ip_pool_id,
+        }
+    }
+}
+
+impl From<&IpPoolRange> for external::IpRange {
+    fn from(range: &IpPoolRange) -> Self {
+        let maybe_range =
+            match (range.first_address.ip(), range.last_address.ip()) {
+                (IpAddr::V4(first), IpAddr::V4(last)) => {
+                    external::IpRange::try_from((first, last))
+                }
+                (IpAddr::V6(first), IpAddr::V6(last)) => {
+                    external::IpRange::try_from((first, last))
+                }
+                (first, last) => {
+                    unreachable!(
```

Comment:
I wonder if we could make this more type-safe (and avoid some boilerplate) with a single `IpRange<IpType>`....but I'm not sure it's worth it.

Reply:
I agree it'd be nice. I played around with it a while ago, but didn't think the approach of creating a marker trait for the IP types was all that great. I'll give it a go again now.

Reply:
I'm going to defer this one.